### ISSUE TITLE
vgl 3.0.90-20221122 crashing torch on ubuntu 22.04

### ISSUE THREAD

**Issue by muratmaga** [VirtualGL/virtualgl#227]
We are using vgl on a cloud server with A100 gpu (grid driver 525.60.13) with 3D Slicer with egl backend.

If we invoke the slicer without vglrun, our torch based application works fine. If we start the Slicer via 
vglrun -d egl ./Slicer 

a little after the torch starts loading the trained model, application crashes. There is no error message (neither in Slicer not in vglrun) that I can locate. 

I am assuming it is a weird combination of the driver being used and VGL, as we can't replicate this crash on our local machine (though we use ubuntu 20.04, not 22.04, and we have RTX A4000 not A100). 

I would appreciate some pointers on how to troubleshoot this. Thanks.

**Comment by dcommander** [comment]
Does it segfault?  If so, can you get a stack trace to determine where it is crashing?  (NOTE: Running the application with `vglrun +de` will cause VirtualGL to pause so you can attach GDB.)

Does Slicer otherwise work with `vglrun -d egl`?  In other words, is it only a specific Slicer workflow (your torch-based application) that fails?  I don't know what "torch" or a "torch-based application" is.

It's been my experience that the GPU hardware rarely matters when diagnosing VirtualGL-related crashes, but the driver version might.  Have you tested the [non-GRID version of 525.60.xx](https://www.nvidia.com/download/driverResults.aspx/196723/en-us/) on the local machine?

Have you tried building VirtualGL from source on the cloud server?  Our official binaries are built on CentOS 7, so perhaps something changed in one of the ABIs that is causing difficulties.

Have you verified that you can run other applications, such as GLXspheres, using `vglrun -d egl` on the cloud server?

**Comment by dcommander** [comment]
It would also be worth testing whether this same crash occurs with VirtualGL 3.0.2.  That will tell me whether it is caused by a new feature in VirtualGL 3.1 beta.

**Comment by muratmaga** [comment]
Torch is a ML framework in python (also called pytorch). https://pytorch.org/

yes, Slicer works completely find with vglrun -d egl (i.e for regular 3D rendering using GPU) and , and all other vglrun sanity checks (glxspheres etc) are fine too. Unfortunately I cannot install the non-grid driver as those are baked into the image the cloud provider provides (I tried once, and everything went haywire). 

I will try the remaining suggestions and report back later.

**Comment by dcommander** [comment]
I suggested that you try the non-GRID driver on the local machine, not on the cloud server. Since you are unable to reproduce the issue on the local machine, perhaps upgrading the driver to a similar version as the driver on the cloud server will cause the issue to occur. This is all in the interest of collecting data points to isolate the issue's cause. If upgrading the driver on the local machine causes the issue to occur, then it is likely that I will also be able to reproduce the issue by upgrading the driver on my machine. If the issue always occurs with VirtualGL 3.1 beta and not with 3.0.2, then it is likely a regression caused by a feature in 3.1 beta, and it is also likely that I will be able to reproduce it. If, however, the issue is likely specific to the cloud server, then there is no sense in me wasting my time trying to repro it locally.

**Comment by dcommander** [comment]
Also, please specify exactly how the application is crashing. Searching the issues, I came across another one related to PyTorch and VGL that has to do with certain libraries not being picked up if they are linked with an RPATH of $ORIGIN.

**Comment by zacharygraber** [comment]
Hi there! @dcommander -- I work with the cloud system in question here ([Jetstream2](https://jetstream-cloud.org/)), and I figured I might be able to jump in and give a little context about the setup. We have a number of A100s in our GPU compute nodes that are provisioned to users through VMs; because not everyone is interested in using an _entire_ A100, though, we use NVIDIA's proprietary GRID drivers to effectively "slice up" an A100 into 5 parts.

For example, when a user spins up a VM, they may choose a small GPU "flavor." In that case, the hypervisor will "slice up" a physical GPU and pass through to the VM a smaller virtual GPU (vGPU); on the backend this vGPU only represents ~1/5th of an A100's compute and VRAM, but the virtual machine effectively sees it as one physical card on a PCI interface. This is why it is impossible to install a non-GRID driver on the cloud machine (VM); NVIDIA's proprietary GRID drivers will recognize the vGPU and understand how to properly use it, while a non-GRID driver is only really intended for communicating with _an entire card_ that's _physically connected_ to the machine. 

Unfortunately, NVIDIA's licensing terms for the GRID drivers and vGPU solutions in general are extremely restrictive. Our sysadmins receive driver installer packages directly from NVIDIA and we are unable to distribute them to anyone else; in other words, we can provide an OS image with the drivers pre-installed, but we cannot give a user the `.deb` or `.rpm` file directly. This makes troubleshooting matters like this a bit difficult, especially if you fear the issue might be in the drivers (for example, the user has no way of setting up the GRID drivers on their local machine).

Hopefully this gives a bit of context. Since the setup on Jetstream2 differs so much from what a local machine would have (one GPU directly hooked into one machine with a "normal" driver), **unless anyone is able to reproduce the issue on their local machine**, comparing functionality on locals versus the cloud VM is a bit of an "apples to oranges" scenario.

**Comment by dcommander** [comment]
@zacharygraber I understand all of that. I've been developing VirtualGL for nearly 20 years, and it is my experience that, when nVidia drivers cause a problem with VirtualGL, it is because of something in nVidia's libGL implementation, which is abstracted from the hardware.  Thus, it might be a useful data point to test a non-GRID installation of 525.60.xx on the local machine.

However, that assertion is coming from a place of having no information regarding how the application is crashing, because @muratmaga has not yet provided me with that information.  It's entirely possible that this is the same issue as https://github.com/VirtualGL/virtualgl/issues/107, in which case it is a known issue with PyTorch that has to be worked around when using that framework with VGL.  (There's even an application recipe for it in the VirtualGL User's Guide: https://rawcdn.githack.com/VirtualGL/virtualgl/3.1beta1/doc/index.html#hd0015.)

In other words, if you guys want more intelligent suggestions from me, then please provide more information about the problem.  Otherwise, I'm shooting in the dark.

**Comment by muratmaga** [comment]
@dcommander I didn't have the time to try some of the more involved things you asked such as trying grid driver on our local machine (which I can't do it as we constantly use it for many tasks) and debugging. But I did try things with vgl 3.0.2, and crash replicates. So it is not something recently introduced with the beta version. 

I also tried the recipe for issue #107 and this is what I get:

**Comment by dcommander** [comment]
If I had any confidence that it would be reproducible on my machine, I would try it, but I can't afford to spend several hours on a wild goose chase.  The best suggestion I have at the moment is to pay my hourly consulting rate and let me log in and diagnose it remotely using a clone of your cloud computing image.

**Comment by dcommander** [comment]
Is there any new information on this issue?

**Comment by muratmaga** [comment]
All I can say, it continues to crash. This is with vgl 3.1-20230315, Nvidia Driver Version: 525.85.05, and torch 2.1.0+cu118

When we need to run models, we start Slicer without the vglrun. 

```
exouser@tvnc:~/Slicer-5.4.0-linux-amd64$ vglrun -d egl ./Slicer 
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: too many profiles
Switch to module:  "Welcome"
Switch to module:  "MEMOS"
"Volume" Reader has successfully read the file "/home/exouser/Downloads/IMPC_sample_data.nrrd" "[0.13s]"
Checking python dependencies
Requirement already satisfied: pillow in ./lib/Python/lib/python3.9/site-packages (10.0.0)

[notice] A new release of pip is available: 23.1.2 -> 23.3.1
[notice] To update, run: python-real -m pip install --upgrade pip
"Color" Reader has successfully read the file "/home/exouser/Slicer-5.4.0-linux-amd64/slicer.org/Extensions-31938/MEMOS/lib/Slicer-5.4/qt-scripted-modules/Resources/Support/KOMP2.ctbl" "[0.01s]"
Using device:  0
MONAI version: 0.9.0
Numpy version: 1.25.1
Pytorch version: 2.1.0+cu118
MONAI flags: HAS_EXT = False, USE_COMPILED = False
MONAI rev id: af0e0e9f757558d144b655c63afcea3a4e0a06f5
MONAI __file__: /home/exouser/Slicer-5.4.0-linux-amd64/lib/Python/lib/python3.9/site-packages/monai/__init__.py

Optional dependencies:
Pytorch Ignite version: NOT INSTALLED or UNKNOWN VERSION.
Nibabel version: 5.1.0
scikit-image version: NOT INSTALLED or UNKNOWN VERSION.
Pillow version: 10.0.0
Tensorboard version: NOT INSTALLED or UNKNOWN VERSION.
gdown version: NOT INSTALLED or UNKNOWN VERSION.
TorchVision version: 0.16.0+cu118
tqdm version: NOT INSTALLED or UNKNOWN VERSION.
lmdb version: NOT INSTALLED or UNKNOWN VERSION.
psutil version: NOT INSTALLED or UNKNOWN VERSION.
pandas version: NOT INSTALLED or UNKNOWN VERSION.
einops version: 0.7.0
transformers version: NOT INSTALLED or UNKNOWN VERSION.
mlflow version: NOT INSTALLED or UNKNOWN VERSION.
pynrrd version: 1.0.0

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies

Using device:  cuda
error: [/home/exouser/Slicer-5.4.0-linux-amd64/bin/SlicerApp-real] exit abnormally - Report the problem.
```

**Comment by muratmaga** [comment]
Slight update. with 535 series driver, we know get an error message instead Slicer exit abnormally. 
```
Unable to load any of {libcudnn_graph.so.9.1.0, libcudnn_graph.so.9.1, libcudnn_graph.so.9, libcudnn_graph.so}
Invalid handle. Cannot load symbol cudnnCreate
error: [/home/exouser/Slicer/bin/./python-real] exit abnormally - Report the problem.
```
This now replicates with 535 drivers provided by the Ubuntu (previously it was specific to the GRID driver). Specifics of the system:

**Comment by dcommander** [comment]
I wonder if it's related to #209.

**Comment by stay-leave** [comment]
> Slight update. with 535 series driver, we now get an error message instead Slicer exit abnormally.
> 
> ```
> Unable to load any of {libcudnn_graph.so.9.1.0, libcudnn_graph.so.9.1, libcudnn_graph.so.9, libcudnn_graph.so}
> Invalid handle. Cannot load symbol cudnnCreate
> error: [/home/exouser/Slicer/bin/./python-real] exit abnormally - Report the problem.
> ```
> 
> Also another change is this used to unique to the grid driver. But now it replicates with 535 drivers provided by the Ubuntu. Specifics of the system:
> 
> Ubuntu 22.04 Nvidia driver version: 535.183.01 vgl version: 3.1.1 cuda version: 11.8 torch version: 2.4.0

Hello, do you slove this? I meet this same error.

**Comment by muratmaga** [comment]
Unfortunately not. Our solution end up not invoking vglrun when we are running torch inside slicer. It is a bit inconvenient but it works.

**Comment by stay-leave** [comment]
> Unfortunately not. Our solution end up not invoking vglrun when we are running torch inside slicer. It is a bit inconvenient but it works.

thanks， i will use cpu. I try all solutions,but still :Unable to load any of {libcudnn_graph.so.9.1.0, libcudnn_graph.so.9.1, libcudnn_graph.so.9, libcudnn_graph.so}

**Comment by dcommander** [comment]
Is there any way I could reproduce this locally, without a container or cloud computing image?

**Comment by muratmaga** [comment]
For me this replicates on our lab ubuntu workstation with regular Nvidia drivers 535 series. 
Setup takes a bit though:

1. Install 3D Slicer (download.slicer.org, use the latest stable 5.8.0). Here is a link to a few libraries you might need to apt install: https://slicer.readthedocs.io/en/latest/user_guide/getting_started.html#linux
2. Install MEMOS extension (blue E icon on the upper right), Slicer will want to restart
4. Download this file https://raw.githubusercontent.com/SlicerMorph/SampleData/master/IMPC_sample_data.nrrd
5. To create the crash, start slicer with vglrun (make sure python console is open for you to see the python outputs, including the error)
6. Load the sample data into Slicer (drag and drop to the application window)
7. Switch to the MEMOS module (CTRL+F search for memos). The initial switch to memos may take a while, since Pytorch module will bring the all necessary python packages. It takes about a minute (or more depending on the connection). 
8. Set the Volume to the sample data, and choose the default segmentation model. 

and wait about 10 second (the point torch starts loading the necessary cuda libraries). 

Execute the same worfklow without vglrun and see that it works fine. 


The complication, with the cloud was that they were using the Nvidia grid driver, which was necessary to partition the GPU into vGPUs. They have changed that. Now, one a full GPU is provided as a passthrough device, and we just use the regular Nvidia ubuntu drivers. The cloud one also crashes with the same error. So if you like, I can create the setup for you on the cloud for you to debug (so that you don't have to do all these steps). But it is your call.

**Comment by dcommander** [comment]
I need to set it up locally, because diagnosis will probably involve instrumenting and rebuilding VirtualGL from source.

**Comment by muratmaga** [comment]
any insights?

**Comment by dcommander** [comment]
> any insights?

It is in the queue, but there's a high-priority TurboVNC bug fix that I have to address first.

**Comment by dcommander** [comment]
I had to add **{slicer-directory}/lib/Python/lib/python3.9/site-packages/nvidia/cudnn/lib** to `LD_LIBRARY_PATH` in order to get the MEMOS extension to start.  Now I'm running into a GPU memory error:

torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 512.00 MiB. GPU 0 has a total capacity of 1.94 GiB of which 405.50 MiB is free. Including non-PyTorch memory, this process has 1.25 GiB memory in use. Of the allocated memory 1.18 GiB is allocated by PyTorch, and 18.43 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

I have tried setting the environment variable as recommended, but that didn't help.  Unfortunately my only nVidia GPU is a Quadro P620, which is somewhat low-end.

**Comment by muratmaga** [comment]
Yeah, the inference model requires a GPU with more than 6GB of RAM and probably will not work with your P620. But that means you are no encountering the dynamic library error, which was blocking for us. So, adding the {slicer-directory}/lib/Python/lib/python3.9/site-packages/nvidia/cudnn/lib to LD_LIBRARY_PATH fixed the issue? Or you never had it?

I will give that a try.

**Comment by muratmaga** [comment]
Yep, that definitely fixes it. After that torch loads the model fine, and we can get the inference. But the question is why is it happening with vglrun. Because I don't have that issue if I don't run Slicer (and MEMOS) without the vglrun.

**Comment by muratmaga** [comment]
funnily enough on the cloud instance with a single GPU I can actually make it work without specifying the LD_LIBRARY_PATH with -nodl option. (vglrun -nodl ~/Slicer/Slicer and then inference completes).
But in our lab computer where I specify the GPU to be used for rendering, I get this error:

maga@magalab-ML:~/Desktop$ vglrun -d /dev/dri/card3 -nodl ~/Downloads/Slicer/Slicer 
[VGL] ERROR: Could not load EGL functions
[VGL]    /lib/libvglfaker-nodl.so: undefined symbol: eglGetProcAddress

It is the same error even if I don't specify the card
maga@magalab-ML:~/Desktop$ vglrun  -nodl ~/Downloads/Slicer/Slicer 
[VGL] ERROR: Could not load EGL functions
[VGL]    /lib/libvglfaker-nodl.so: undefined symbol: eglGetProcAddress

**Comment by dcommander** [comment]
Same issue as #107.  The PyTorch dynamic libraries are linked with an RPATH of `$ORIGIN`.  Normally that causes the dynamic linker to look for dependencies (including libraries opened explicitly with `dlopen()`) in the same directory as the dynamic library.  However, when VirtualGL's `dlopen()` interposer is active, it causes the dynamic linker to look for dependencies in the same directory as **libdlfaker.so**.  Thus, you have to explicitly add the dependency directory to `LD_LIBRARY_PATH` or pass it to vglrun using the `-ld` option.  (This is in fact a documented recipe for PyTorch in the VirtualGL User's Guide.)  The missing piece of the puzzle was the need to add the Cuda library directories rather than the PyTorch library directories to `LD_LIBRARY_PATH`.  That's why the recipe didn't work for you in https://github.com/VirtualGL/virtualgl/issues/227#issuecomment-1447474534.  I'll look into expanding the recipe to be more clear that it applies to dependencies as well.

Passing `-nodl` to vglrun disables the `dlopen()` interposer, which works around the RPATH issue.  However, that won't work if the application uses `dlopen()` to open any OpenGL-related library.  Not sure why it works on one system and not the other, but that could be explained by differing driver versions.

**Comment by dcommander** [comment]
Another note:

In my case, PyTorch did produce an error message that clued me in to the necessary `LD_LIBRARY_PATH` modifications:

Unable to load any of {libcudnn_graph.so.9.1.0, libcudnn_graph.so.9.1, libcudnn_graph.so.9, libcudnn_graph.so}
Invalid handle. Cannot load symbol cudnnCreate

I am running Rocky Linux 8 rather than Ubuntu, so the Python infrastructure on my system is different than yours.

**Comment by dcommander** [comment]
#276 has been fixed, so the latest pre-release builds should work automatically with 3D Slicer and `-nodl`.

**Comment by muratmaga** [comment]
Works like a charm thank you for fixing it.
